src/dotnet/commands/dotnet-build/ProjectBuilder.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections.Generic;
+using NuGet.Frameworks;
+using Microsoft.DotNet.ProjectModel;
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    internal enum CompilationResult
+    {
+        IncrementalSkip, Success, Failure
+    }
+
+    internal abstract class ProjectBuilder
+    {
+        private readonly bool _skipDependencies;
+
+        public ProjectBuilder(bool skipDependencies)
+        {
+            _skipDependencies = skipDependencies;
+        }
+
+        private Dictionary<ProjectContextIdentity, CompilationResult> _compilationResults = new Dictionary<ProjectContextIdentity, CompilationResult>();
+
+        public IEnumerable<CompilationResult> Build(IEnumerable<ProjectGraphNode> roots)
+        {
+            foreach (var projectNode in roots)
+            {
+                yield return Build(projectNode);
+            }
+        }
+
+        private CompilationResult Build(ProjectGraphNode projectNode)
+        {
+            var projectContextIdentity = new ProjectContextIdentity(projectNode.ProjectContext);
+
+            CompilationResult result;
+            if (_compilationResults.TryGetValue(projectContextIdentity, out result))
+            {
+                return result;
+            }
+            result = CompileWithDependencies(projectNode);
+
+            _compilationResults[projectContextIdentity] = result;
+
+            return result;
+        }
+
+        private CompilationResult CompileWithDependencies(ProjectGraphNode projectNode)
+        {
+            var skipIncrementalCheck = false;
+            if (!_skipDependencies)
+            {
+                foreach (var dependency in projectNode.Dependencies)
+                {
+                    var result = Build(dependency);
+                    switch (result)
+                    {
+                        case CompilationResult.IncrementalSkip:
+                            break;
+                        case CompilationResult.Success:
+                            skipIncrementalCheck = false;
+                            break;
+                        case CompilationResult.Failure:
+                            return CompilationResult.Failure;
+                        default:
+                            throw new NotSupportedException();
+                    }
+                }
+            }
+            if (SkipIncremental(projectNode))
+            {
+                return CompilationResult.IncrementalSkip;
+            }
+            else
+            {
+                return RunCompile(projectNode);
+            }
+        }
+
+        protected abstract CompilationResult RunCompile(ProjectGraphNode projectNode);
+
+        protected abstract bool SkipIncremental(ProjectGraphNode projectNode);
```

Comment:
I'd call this something like "PreviousBuildIsNotOutOfDateAndCanBeReusedWithoutRecompilingTheProject"? :trollface: 

Troll names aside, it wasn't exactly clear what this method was supposed to do until I reviewed the code in `DotNetProjectBuilder`. It seemed like maybe it was supposed to return true if we skip the incremental build? But I don't even 100% know what that means :P. Maybe just call it "NeedsRebuilding" and merge it with the existing method on DotNetProjectBuilder? Or "IsUpToDate" or something?